### PR TITLE
support trino query paimon file index for map type with map-keys

### PR DESCRIPTION
**Pull request by zyl891229**
support trino query paimon file index for map type with map-keys

enable map type  index as follow 

```
CREATE TABLE my_table (age int , jsonmap MAP<STRING,STRING> ) WITH
(
"file-index.bloom-filter.columns" = "jsonmap[a]", 
)
```

query map type  index as follow  inpaimon trino must use fun element_at to extract value

`select * from paimon.default.my_table where age = 30 and element_at(jsonmap, 'a')  = '1';`

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -348,6 +348,19 @@ under the License.
...
             </plugin>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-source-plugin</artifactId>
+                <version>3.2.1</version>
+                <executions>
+                    <execution>
+                        <id>attach-sources</id>
+                        <goals>
+                            <goal>jar-no-fork</goal>
```
Comment (leaves12138) [inline pom.xml:359]: Can you explain why use this goal "jar-no-fork"

File: paimon-trino-common/src/main/java/org/apache/paimon/trino/TrinoFilterConverter.java [hunk 2]
```diff
@@ -69,6 +72,7 @@
 public class TrinoFilterConverter {
 
     private static final Logger LOG = LoggerFactory.getLogger(TrinoFilterConverter.class);
+    public static final Pattern IS_NESTED = Pattern.compile(".+\\[.+]");
```
Comment (leaves12138) [inline paimon-trino-common/src/main/java/org/apache/paimon/trino/TrinoFilterConverter.java:75]: Please refer to https://github.com/apache/paimon/pull/3221, it does not use regex anymore

File: paimon-trino-common/src/main/java/org/apache/paimon/trino/TrinoFilterConverter.java [hunk 3]
```diff
@@ -103,10 +107,20 @@ public Optional<Predicate> convert(
...
             int index = fieldNames.indexOf(field);
             if (index != -1) {
                 try {
-                    conjuncts.add(toPredicate(index, columnHandle.getTrinoType(), domain));
+                    conjuncts.add(
+                            toPredicate(
+                                    index,
+                                    columnHandle.getColumnName(),
+                                    columnHandle.getTrinoType(),
+                                    domain));
```
Comment (leaves12138) [inline paimon-trino-common/src/main/java/org/apache/paimon/trino/TrinoFilterConverter.java:123]: We should pass the nested field to the predicate. For example, new LeafPredicate(columnName="map_data[key1]")

File: paimon-trino-common/src/main/java/org/apache/paimon/trino/TrinoMetadata.java [hunk 4]
```diff
@@ -448,7 +461,37 @@ public Optional<ConstraintApplicationResult<ConnectorTableHandle>> applyFilter(
...
 
         return Optional.of(
-                new ConstraintApplicationResult<>(trinoTableHandle.copy(newFilter), remain, false));
+                new ConstraintApplicationResult<>(
+                        trinoTableHandle.copy(newFilter, expressionFilter), remain, false));
+    }
+
+    private static TupleDomain<TrinoColumnHandle> getTrinoColumnHandleTupleDomain(
+            Constraint constraint) {
+        Map<TrinoColumnHandle, Domain> expressionPredicates = new HashMap<>();
```
Comment (leaves12138) [inline paimon-trino-common/src/main/java/org/apache/paimon/trino/TrinoMetadata.java:470]: I think we could just declair expressionPredicate. In default: we put expressionPredicates=Collections.emptyMap()

File: paimon-trino-common/src/main/java/org/apache/paimon/trino/TrinoMetadata.java [hunk 5]
```diff
@@ -448,7 +461,37 @@ public Optional<ConstraintApplicationResult<ConnectorTableHandle>> applyFilter(
...
+
+        if (constraint.getExpression() instanceof Call) {
+            Call expression = (Call) constraint.getExpression();
+            Map<String, ColumnHandle> assignments = constraint.getAssignments();
+
+            switch (expression.getFunctionName().getName()) {
+                case "$equal":
```
Comment (leaves12138) [inline paimon-trino-common/src/main/java/org/apache/paimon/trino/TrinoMetadata.java:477]: Can we put there in static final string?

private static final String TRINO_EQUAL = "$equal"

File: paimon-trino-common/src/main/java/org/apache/paimon/trino/TrinoMetadata.java [hunk 6]
```diff
@@ -448,7 +461,37 @@ public Optional<ConstraintApplicationResult<ConnectorTableHandle>> applyFilter(
...
+                    expressionPredicates = handleElementAtArguments(assignments, expression, false);
+                    break;
+
+                case "$in":
```
Comment (leaves12138) [inline paimon-trino-common/src/main/java/org/apache/paimon/trino/TrinoMetadata.java:481]: ditto

File: paimon-trino-common/src/main/java/org/apache/paimon/trino/TrinoMetadata.java [hunk 7]
```diff
@@ -448,7 +461,37 @@ public Optional<ConstraintApplicationResult<ConnectorTableHandle>> applyFilter(
...
+                    expressionPredicates = handleElementAtArguments(assignments, expression, true);
+                    break;
+
+                case "$and":
```
Comment (leaves12138) [inline paimon-trino-common/src/main/java/org/apache/paimon/trino/TrinoMetadata.java:485]: ditto

File: paimon-trino-common/src/main/java/org/apache/paimon/trino/TrinoMetadata.java [hunk 8]
```diff
@@ -516,4 +559,83 @@ public Optional<LimitApplicationResult<ConnectorTableHandle>> applyLimit(
...
         return Optional.of(new LimitApplicationResult<>(table, false, false));
     }
+
+    private static Map<TrinoColumnHandle, Domain> handleElementAtArguments(
+            Map<String, ColumnHandle> assignments, Call expression, boolean isIn) {
+        Map<TrinoColumnHandle, Domain> expressionPredicates = new HashMap<>();
+
+        Call elementAtExpression = (Call) expression.getArguments().get(0);
+
+        if (!elementAtExpression.getFunctionName().getName().equals("element_at")) {
```
Comment (leaves12138) [inline paimon-trino-common/src/main/java/org/apache/paimon/trino/TrinoMetadata.java:569]: ditto, element_at is a special function name, put it in static final String

File: paimon-trino-common/src/main/java/org/apache/paimon/trino/TrinoMetadata.java [hunk 9]
```diff
@@ -516,4 +559,83 @@ public Optional<LimitApplicationResult<ConnectorTableHandle>> applyLimit(
 
         return Optional.of(new LimitApplicationResult<>(table, false, false));
     }
+
+    private static Map<TrinoColumnHandle, Domain> handleElementAtArguments(
```
Comment (leaves12138) [inline paimon-trino-common/src/main/java/org/apache/paimon/trino/TrinoMetadata.java:563]: I think we can't put these expression parser code in a separated class, just left TrinoMetadata clean.

We can make a new class parse from expression to paimon Predicate

Comment (leaves12138) [reply paimon-trino-common/src/main/java/org/apache/paimon/trino/TrinoMetadata.java:563]: And we can add some UT test for this new class if possible.

File: paimon-trino-common/src/main/java/org/apache/paimon/trino/TrinoExpressionFilterExtract.java [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+import java.util.HashMap;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import static io.trino.spi.expression.StandardFunctions.AND_FUNCTION_NAME;
+import static io.trino.spi.expression.StandardFunctions.EQUAL_OPERATOR_FUNCTION_NAME;
+import static io.trino.spi.expression.StandardFunctions.IN_PREDICATE_FUNCTION_NAME;
+
+/** Extract Trino Expression filter ( e.g. element_at(jsonmap, 'a') = '1' ) to TrinoColumnHandle. */
+public class TrinoExpressionFilterExtract {
```
Comment (JingsongLi) [inline paimon-trino-common/src/main/java/org/apache/paimon/trino/TrinoExpressionFilterExtract.java:48]: `TrinoMapKeyFilterExtract`?

File: paimon-trino-common/src/main/java/org/apache/paimon/trino/TrinoMetadata.java [hunk 11]
```diff
@@ -423,6 +424,10 @@ public Optional<ConstraintApplicationResult<ConnectorTableHandle>> applyFilter(
                         .getSummary()
```
Comment (JingsongLi) [inline paimon-trino-common/src/main/java/org/apache/paimon/trino/TrinoMetadata.java:424]: add comment here: what is returned from `getSummary`.

File: paimon-trino-common/src/main/java/org/apache/paimon/trino/TrinoExpressionFilterExtract.java [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+                                                    Range.equal(
+                                                            elementAtValue.getType(),
+                                                            elementAtValue.getValue()))),
+                            false));
+        }
+        return expressionPredicates;
+    }
+
+    public static String toMapKey(String mapColumnName, String keyName) {
+        return mapColumnName + "[" + keyName + "]";
```
Comment (JingsongLi) [inline paimon-trino-common/src/main/java/org/apache/paimon/trino/TrinoExpressionFilterExtract.java:121]: add comment: Follow Paimon API definition.

**Comment by leaves12138** [comment]
+1

**Pull request merged**